README.md

```diff
@@ -216,34 +215,109 @@ If you register the service providers in the correct order in your container (A
+Now, let's have a look at this similar code:
 
-Using static methods *conveys* the idea that everything should be stateless. Furthermore, since factories are supposed to be stateless, there is no need for a factory to rely on the service provider state (using `$this`), hence the idea to make factories static.
+```php
+class MyServiceProvider implements ServiceProvider
+{
+    public function getServices()
+    {
+        return [
+            'my_service' => [ MyServiceProvider::class, 'createMyService' ],
+        ];
+    }
+    
+    public static function createMyService()
+    {
+        return new MyService();
+    }
+}
+```
 
-Also static methods are easy to call, no need to instantiate any object which, when building a container in every request, is a bit of a performance advantage compared to having to create new objects for no reason.
+Compiled/cached container can now run `getServices()` once and subsequently directly refer to the factory method (`MyServiceProvider::createMyService`).
 
-#### Why does the `getServices()` method returns an array of function names instead of an array of callables?
+Now, the minimal code to fetch the service becomes:
 
-An array of callables means anonymous functions could be returned, which is hard to adapt to compiled or cached containers.
+```php
+$service = MyServiceProvider::createMyService()();
+```
 
-Even if we explicitly forbid closures (which would be weird in the first place), invokable objects or object methods would also hard to adapt with compiled/cached containers.
+So the second solution offers more room for optimization for compiled/cached containers.
 
-So the only callables that would really be allowed would be:
+As a courtesy for users looking for the best performance, service providers may consider using `public static function`s as factories.
 
-- function name
-- static method (['Class', 'method'])
+### Managing configuration
 
-It seems a bit weird to use function for this use case (especially since they are not autoloadable). So we end up with static methods.
+The service created by a factory should only depend on the input parameters of the factory (`$container` and `$getPrevious`).
+If the factory needs to fetch parameters, those should be fetched from the container directly.
 
-Of course, we could allow the static methods to be in another class than the service provider, but it seems generally a good idea to have the services created in the service provider.
+```php
+class MyServiceProvider implements ServiceProvider
+{
+    public function getServices()
+    {
+        return [
+            'logger' => [ MyServiceProvider::class, 'createLogger' ],
+        ];
+    }
+    
+    public static function createLogger(ContainerInterface $container)
+    {
+        // The path to the log file is fetched from the container, not from the service provider state.
+        return new FileLogger($this->container->get('logFilePath'));
+    }
+}
+```
```

Comment:
Just realized that this paragraph was moved actually: I don't necessarily think it needs to be expanded.